crates/chain/src/tx_graph.rs

```diff
@@ -52,63 +54,124 @@
 //! let additions = graph.apply_update(update);
 //! assert!(additions.is_empty());
 //! ```
-use crate::{collections::*, ForEachTxOut};
+
+use crate::{collections::*, BlockAnchor, ChainOracle, ForEachTxOut, ObservedIn};
 use alloc::vec::Vec;
 use bitcoin::{OutPoint, Transaction, TxOut, Txid};
-use core::ops::RangeInclusive;
+use core::{
+    convert::Infallible,
+    ops::{Deref, RangeInclusive},
+};
 
 /// A graph of transactions and spends.
 ///
 /// See the [module-level documentation] for more.
 ///
 /// [module-level documentation]: crate::tx_graph
-#[derive(Clone, Debug, PartialEq, Default)]
-pub struct TxGraph {
-    txs: HashMap<Txid, TxNode>,
+#[derive(Clone, Debug, PartialEq)]
+pub struct TxGraph<A = ()> {
+    // all transactions that the graph is aware of in format: `(tx_node, tx_anchors, tx_last_seen)`
+    txs: HashMap<Txid, (TxNodeInternal, BTreeSet<A>, u64)>,
     spends: BTreeMap<OutPoint, HashSet<Txid>>,
+    anchors: BTreeSet<(A, Txid)>,
 
     // This atrocity exists so that `TxGraph::outspends()` can return a reference.
     // FIXME: This can be removed once `HashSet::new` is a const fn.
     empty_outspends: HashSet<Txid>,
 }
 
-/// Node of a [`TxGraph`]. This can either be a whole transaction, or a partial transaction (where
-/// we only have select outputs).
+impl<A> Default for TxGraph<A> {
+    fn default() -> Self {
+        Self {
+            txs: Default::default(),
+            spends: Default::default(),
+            anchors: Default::default(),
+            empty_outspends: Default::default(),
+        }
+    }
+}
+
+// pub type InChainTx<'a, T, A> = (ObservedIn<&'a A>, TxInGraph<'a, T, A>);
+// pub type InChainTxOut<'a, I, A> = (&'a I, FullTxOut<ObservedIn<&'a A>>);
+
+/// An outward-facing view of a transaction node that resides in a [`TxGraph`].
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TxNode<'a, T, A> {
```

Comment:
I am finding the naming more confusing than before. This is not a node in the graph. And the doc above says it's an "outward-facing view". Maybe it could be something like `TxGraphData`, or simply `TxData`. 

A bit of simplification of the doc above would also help. 
`A transaction with extra metadata that is in the graph` or something like that.

Reply:
The graph contains the structure `TxNodeInternal` which was previously `TxNode`, so that makes more sense to me as a "node". And this is a transaction data with extra metadata. Anyway, if the intended purpose of the structure is clear, then no issues. The Doc could be a bit more clear too.